### ISSUE TITLE
Compiler Crash on db_migrate 

### ISSUE THREAD

**Issue by sdogruyol** [ysbaddaden/frost#7]
On OS X 10.10 (El Capitan) I successfully run the app with `make run`. 

But when i try to migrate following the frost guide the compiler crashed with this stacktrace.

```
/usr/local/bin/crystal run bin/db -- migrate
Bug: `def ->(a : CreatePosts, b : CreatePosts)
  a.version
end` at  has no type
[4315581051] *raise<String>:NoReturn +139
[4316482424] *Crystal::Def@Crystal::ASTNode#type<Crystal::Def>:Crystal::Type+ +488
[4319091402] *Crystal::CodeGenVisitor#codegen_fun_signature_non_external<Crystal::CodeGenVisitor, String, Crystal::Def, Crystal::Type+, Bool, Bool>:Array(Crystal::Arg) +1018
[4319087632] *Crystal::CodeGenVisitor#codegen_fun<Crystal::CodeGenVisitor, String, Crystal::Def, Crystal::Type+, Bool, LLVM::Module, Bool, Bool>:LLVM::Function +624
[4319039265] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::ASTNode+, Crystal::CodeGenVisitor>:Nil +3569
[4319144274] *Crystal::CodeGenVisitor#visit<Crystal::CodeGenVisitor, Crystal::Call>:Bool +562
[4319036814] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::ASTNode+, Crystal::CodeGenVisitor>:Nil +1118
[4319124964] *Crystal::CodeGenVisitor#visit<Crystal::CodeGenVisitor, Crystal::Assign>:(Nil | Bool) +324
[4319037945] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::ASTNode+, Crystal::CodeGenVisitor>:Nil +2249
[4319129873] *Crystal::CodeGenVisitor#codegen_if_branch<Crystal::CodeGenVisitor, Crystal::CodeGenVisitor::Phi, Crystal::ASTNode+, LLVM::BasicBlock, Bool>:(Nil | LLVM::Value) +65
[4319048171] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::ASTNode+, Crystal::CodeGenVisitor>:Nil +12475
[4319113986] *Crystal::CodeGenVisitor#codegen_fun<Crystal::CodeGenVisitor, String, Crystal::Def+, Crystal::Type+, Bool, LLVM::Module, Bool, Bool>:LLVM::Function +3314
[4319103953] *Crystal::CodeGenVisitor#target_def_fun<Crystal::CodeGenVisitor, Crystal::Def+, Crystal::Type+>:LLVM::Function +4705
[4319146877] *Crystal::CodeGenVisitor#visit<Crystal::CodeGenVisitor, Crystal::Call>:Bool +3165
[4319036814] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::ASTNode+, Crystal::CodeGenVisitor>:Nil +1118
[4319162931] *Crystal::CodeGenVisitor#prepare_call_args_non_external<Crystal::CodeGenVisitor, Crystal::Call, Crystal::Def+, Crystal::Type+>:{Array(LLVM::Value), Bool} +211
[4319144189] *Crystal::CodeGenVisitor#visit<Crystal::CodeGenVisitor, Crystal::Call>:Bool +477
[4319036814] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::ASTNode+, Crystal::CodeGenVisitor>:Nil +1118
[4319037117] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::ASTNode+, Crystal::CodeGenVisitor>:Nil +1421
[4319129873] *Crystal::CodeGenVisitor#codegen_if_branch<Crystal::CodeGenVisitor, Crystal::CodeGenVisitor::Phi, Crystal::ASTNode+, LLVM::BasicBlock, Bool>:(Nil | LLVM::Value) +65
[4319048142] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::ASTNode+, Crystal::CodeGenVisitor>:Nil +12446
[4319037117] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::ASTNode+, Crystal::CodeGenVisitor>:Nil +1421
[4319089997] *Crystal::CodeGenVisitor#codegen_fun<Crystal::CodeGenVisitor, String, Crystal::Def, Crystal::Type+, Bool, LLVM::Module, Bool, Bool>:LLVM::Function +2989
[4319039265] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::ASTNode+, Crystal::CodeGenVisitor>:Nil +3569
[4319144274] *Crystal::CodeGenVisitor#visit<Crystal::CodeGenVisitor, Crystal::Call>:Bool +562
[4319036814] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::ASTNode+, Crystal::CodeGenVisitor>:Nil +1118
[4318003949] *Crystal::Compiler#compile<Crystal::Compiler, Array(Crystal::Compiler::Source), String>:Crystal::Compiler::Result +4509
[4322128017] *Crystal::Command#run_command<Crystal::Command>:Nil +145
[4321951695] *Crystal::Command#run<Crystal::Command>:(Nil | Bool | Array(String) | IO::FileDescriptor+ | Crystal::Compiler::Result | Array(Crystal::ImplementationTrace) | Array(Crystal::Init::View+:Class)) +3967
[4315579622] __crystal_main +24742
[4321946587] main +43

Error: you've found a bug in the Crystal compiler. Please open an issue, including source code that will allow us to reproduce the bug: https://github.com/manastech/crystal/issues
make: *** [db_migrate] Error 1
```

**Comment by ysbaddaden** [comment]
I overlooked an important snippet `set_version {{ __FILE__ }}` to extract the version number from the filename into the migration class...

I didn't try out manually the code I wrote in the guides (shame on me). I just did test it all, and fixed a few problems in 34d2a6d8f4d6cca70bd056cce0f7d7034c64edbb

**Comment by sdogruyol** [comment]
Great job :+1:
